core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1041,26 +1059,37 @@ public void performOn(ShardRouting replica, ReplicaRequest request, ActionListen
                 return;
             }
             final ConcreteShardRequest<ReplicaRequest> concreteShardRequest =
-                new ConcreteShardRequest<>(request, replica.allocationId().getId());
-            sendReplicaRequest(concreteShardRequest, node, listener);
+                    new ConcreteShardRequest<>(request, replica.allocationId().getId());
+            replicationAction.sendReplicaRequest(concreteShardRequest, node, listener);
+        }
+
+        @Override
+        public boolean failShardIfNeeded(ShardRouting replica, long primaryTerm, String message, Exception exception,
+                                         Runnable onSuccess, Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
+            // This does not need to fail the operation. The idea is that this
+            // is a non-write operation (something like a refresh or a global
+            // checkpoint sync) and therefore the replica should still be
+            // "alive" if it were to fail.
+            return false;
         }
 
         @Override
-        public void failShard(ShardRouting replica, long primaryTerm, String message, Exception exception,
-                              Runnable onSuccess, Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
-            shardStateAction.remoteShardFailed(replica.shardId(), replica.allocationId().getId(), primaryTerm, message, exception,
-                createListener(onSuccess, onPrimaryDemoted, onIgnoredFailure));
+        public void markShardCopyAsStaleIfNeeded(ShardId shardId, String allocationId, long primaryTerm, Runnable onSuccess,
+                                                 Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
+            // Just like with failing the shard, it should only be marked as
+            // stale when write operations fail, so for regular operations
+            // don't mark it as stale when not necessary.
         }
 
         @Override
         public void markShardCopyAsStale(ShardId shardId, String allocationId, long primaryTerm, Runnable onSuccess,
                                          Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
             shardStateAction.remoteShardFailed(shardId, allocationId, primaryTerm, "mark copy as stale", null,
```

Comment:
As I mentioned in my previous comment, we still need a way to mark the replica as stale when the shard is unavailable.

What we don't need is the "if needed" version, because we only do this when the shard is not available, not when an operation fails.

I've removed the "if needed" version of this in 1f58744362b756bc9ef6f6bfb412c2836890d8f7